src/runtime/cuda/nccl_api.cu

```diff
@@ -0,0 +1,832 @@
+  cudaEvent_t d2h;
+  cudaEventCreate(&d2h);
+  cudaEventRecord(d2h, stream);
+
+  // allocate output space
+  cudaEventSynchronize(d2h);
+  cudaEventDestroy(d2h);
+
+  IdArray recv_idx = aten::NewIdArray(recv_prefix_host.back(), ctx, sizeof(IdType)*8);
+
+  std::vector<int64_t> value_shape(in_value->ndim, 0);
+  value_shape[0] = recv_prefix_host.back();
+  for (int d = 1; d < in_value->ndim; ++d) {
+    value_shape[d] = in_value->shape[d];
+  }
+  NDArray recv_value = NDArray::Empty(value_shape, in_value->dtype, ctx);
+
+  // send data
+  comm->SparseAllToAll(
+      send_idx,
+      send_value,
+      num_feat,
+      send_prefix_host.data(),
+      static_cast<IdType*>(recv_idx->data),
+      static_cast<DType*>(recv_value->data),
+      recv_prefix_host.data(),
+      stream);
+  device->FreeWorkspace(ctx, send_idx);
+  device->FreeWorkspace(ctx, send_value);
+
+  return std::pair<IdArray, NDArray>(recv_idx, recv_value);
+}
+
+template<typename IdType, typename DType>
+NDArray SparsePull(
+    NCCLCommunicatorRef comm,
+    IdArray req_idx,
+    NDArray local_tensor,
+    NDArrayPartitionRef part) {
+  const auto& ctx = req_idx->ctx;
+  CHECK_EQ(ctx, local_tensor->ctx);
+  auto device = DeviceAPI::Get(ctx);
+
+  // TODO(dlasalle): Get the stream from the device context.
+  cudaStream_t stream = 0;
+
+  CHECK_EQ(req_idx->ndim, 1);
+
+  const int64_t num_in = req_idx->shape[0];
+  int64_t num_feat = 1;
+  for (int d = 1; d < local_tensor->ndim; ++d) {
+    num_feat *= local_tensor->shape[d];
+  }
+
+  const int64_t comm_size = comm->size();
+
+  if (comm_size == 1) {
+    // Just return index selection from current local_tensor
+    return aten::IndexSelect(local_tensor, req_idx);
+  }
+
+  // First we need to send our requests to other processors. This means
+  // re-ordering our index array to be contiguous among processors, and
+  // counting the number of indices we are sending each processor. For now,
+  // we assume a poorly partitioned graph, and that there exists the
+  // possibility that each processor could request data from this one.
+
+  // the buffer for us to re-order our requests in
+  IdType * send_idx = static_cast<IdType*>(device->AllocWorkspace(ctx,
+      num_in*sizeof(IdType)));
+
+  std::pair<IdArray, NDArray> part_perm = part->GeneratePermutation(req_idx);
+  const IdType * const perm = static_cast<const IdType*>(part_perm.first->data);
+  const int64_t * const send_sum =
+      static_cast<const int64_t*>(part_perm.second->data);
+
+  // permute requests
+  {
+    const dim3 block(256);
+    const dim3 grid((num_in+block.x-1)/block.x);
+
+    aten::impl::IndexSelectSingleKernel<<<grid, block, 0, stream>>>(
+        static_cast<const IdType*>(req_idx->data),
+        perm,
+        num_in,
+        send_idx);
+    CUDA_CALL(cudaGetLastError());
+  }
+
+  // compute the prefix sum of the indexes this process is requesting
+  int64_t * request_prefix = static_cast<int64_t*>(
```

Comment:
Where to release request_prefix?

Reply:
I switched all workspace allocations to use RAII via https://github.com/dmlc/dgl/pull/2825/files#diff-128b2340cfe395c91243eab36571e6f38af6b4eb0d0c3e01eaaba552995d8828 in 
69daf52, to prevent leaking workspace.